### PR TITLE
feat: add last updated time to data cut and master dataset pages

### PR DESCRIPTION
**Pull request by niross**
Closes: https://trello.com/c/U2f069TH/779-spike-show-last-updated-date-and-time-against-data-cuts-and-master-datasets

- Query the new data flow metadata table for master dataset and query last update date
- Query s3 for data cut source link last updated date
 
#### Data cut page before
before

#### Data cut page after
after

### Checklist

* [x] Have tests been added to cover any changes?
* [x] Has the [CHANGELOG](https://github.com/uktrade/data-workspace/blob/master/CHANGELOG.md) been updated?

### CODE REVIEW

File: dataworkspace/dataworkspace/apps/datasets/models.py [hunk 1]
```diff
@@ -470,6 +470,15 @@ def can_show_link_for_user(self, user):
     def type(self):
         return DataLinkType.SOURCE_TABLE.value
 
+    def get_data_last_updated_date(self):
+        from dataworkspace.datasets_db import (
+            get_tables_last_updated_date,
+        )  # pylint: disable=import-outside-toplevel
```
Comment (samuelhwilliams) [inline dataworkspace/dataworkspace/apps/datasets/models.py:476]: Do we need to do this import here rather than just at the top level for some reason?

Comment (niross) [reply dataworkspace/dataworkspace/apps/datasets/models.py:476]: I'm (still) having a bit of a nightmare with cyclic imports it turns out. I may have to just move this to the utils file with the rest of the code that accesses the db.

File: dataworkspace/dataworkspace/datasets_db.py [hunk 2]
```diff
@@ -33,3 +34,32 @@ def query_columns(connection, source):
...
+            FROM (
+                SELECT
+                    CONCAT(metadata.table_schema, '.', metadata.table_name),
+                    MAX(dataflow_swapped_tables_utc) run_date
+                FROM dataflow.metadata
+                WHERE CONCAT(metadata.table_schema, '.', metadata.table_name) = ANY(%s)
+                GROUP BY 1
+            ) a
+            ''',
+            [tables],
```
Comment (samuelhwilliams) [inline dataworkspace/dataworkspace/datasets_db.py:63]: I don't think this query is quite right. I can guess why you're reading `dataflow_swapped_tables_utc`, because the `source_data_modified_utc` column is often NULL - but that's sort of intentional. `dataflow_swapped_tables_utc` doesn't describe when the source data is from - it just says when we made it available on Data Workspace, so it isn't semantically the correct thing to pull/display. We should definitely use `source_data_modified_utc` if it exists. This query as it stands will lie about the ONS datasets in particular - we might run the pipeline every day but the data is only updated once a month/quarter. This would report the ONS data as being updated daily.

We only have values for `source_data_modified_utc` in ONS data right now, I think, because they're the only datasets where we've defined what that value should be. For the remaining pipelines/master datasets we need to make decisions on what that should be and how we get it. For real-time APIs it's not going to be very wrong to use `dataflow_swapped_tables_utc`, but if we have any other pipelines that aren't real time then we aren't going to be showing the right thing - and it may vary on just how wrong it is.

File: dataworkspace/dataworkspace/apps/datasets/models.py (hunk 1 above)
Comment (samuelhwilliams) [reply dataworkspace/dataworkspace/apps/datasets/models.py:476]: Ahhh gotcha, ok. 👍

File: dataworkspace/dataworkspace/datasets_db.py (hunk 2 above)
Comment (niross) [reply dataworkspace/dataworkspace/datasets_db.py]: Got it, i've swapped the table name out

File: dataworkspace/dataworkspace/apps/datasets/models.py (hunk 1 above)
Comment (niross) [reply dataworkspace/dataworkspace/apps/datasets/models.py]: Ok sorted now. I've fixed my local pylint issue as well :+1:

File: dataworkspace/dataworkspace/datasets_db.py [hunk 3]
```diff
@@ -18,18 +17,39 @@ def get_columns(database_name, schema=None, table=None, query=None):
...
+                    CONCAT(metadata.table_schema, '.', metadata.table_name),
+                    MAX(source_data_modified_utc) run_date
+                FROM dataflow.metadata
+                WHERE CONCAT(metadata.table_schema, '.', metadata.table_name) = ANY(%s)
+                GROUP BY 1
+            ) a
+            ''',
+            [tables],
+        )
+        return cursor.fetchone()[0]
```
Comment (samuelhwilliams) [inline dataworkspace/dataworkspace/datasets_db.py:55]: Would this error out if no tables match? I think we might need to support the case where there isn't a known last updated date for a set of tables.

Comment (niross) [reply dataworkspace/dataworkspace/datasets_db.py:55]: No, it just returns None if no tables match. I've added another test for null last updated date as well.

**Review by samuelhwilliams — APPROVED** [review]
Nice 👍 

I wonder if we might prefer to not have a column for "Data last updated" on master datasets when it's not available, but I'm fairly open to being transparent about not having that info yet either.

File: dataworkspace/dataworkspace/apps/datasets/models.py [hunk 4]
```diff
@@ -622,6 +639,14 @@ def parsed_query_tables(self):
         statements = psqlparse.parse(self.query)
         return list(statements[0].tables())
 
+    def get_data_last_updated_date(self):
+        # Ensure all tables have a schema as `public` is not always specified in queries
+        tables = [
+            f'{"public." if len(table_name.split(".")) == 1 else ""}{table_name}'
+            for table_name in self.parsed_query_tables
```
Comment (samuelhwilliams) [inline dataworkspace/dataworkspace/apps/datasets/models.py:646]: I wonder a bit if this might break if the schema/table names have dots or quotes in them (some schemas are definitely dotted). Is this definitely safe? Should we quote/escape them to be extra sure?

**Comment by niross** [comment]
> Nice 
> 
> I wonder if we might prefer to not have a column for "Data last updated" on master datasets when it's not available, but I'm fairly open to being transparent about not having that info yet either.

Hmm yeah, there is a couple of master datasets out there with more than one table though (export wins is the one that comes to mind), so there is a very small chance that one table on the page could have a last updated date and another one doesn't.

**Comment by niross** [comment]
Also... I am going to hold off on merging this temporarily until @abbas123456's query table changes are in.

File: dataworkspace/dataworkspace/apps/datasets/models.py (hunk 4 above)
Comment (niross) [reply dataworkspace/dataworkspace/apps/datasets/models.py:646]: Aha yeah just added another comment. I think this will be unnecessary once https://github.com/uktrade/data-workspace/pull/946 is merged - so I'll hold off until then.

File: dataworkspace/dataworkspace/datasets_db.py [hunk 5]
```diff
@@ -18,18 +17,39 @@ def get_columns(database_name, schema=None, table=None, query=None):
...
+
+        cursor.execute(
+            '''
+            SELECT MIN(run_date)
+            FROM (
+                SELECT
+                    CONCAT(metadata.table_schema, '.', metadata.table_name),
+                    MAX(source_data_modified_utc) run_date
+                FROM dataflow.metadata
+                WHERE CONCAT(metadata.table_schema, '.', metadata.table_name) = ANY(%s)
```
Comment (michalc) [inline dataworkspace/dataworkspace/datasets_db.py:49]: Would it be better/clearer to keep the table and schema names separate in the query, not concatenating them?

- Any indexes on metadata won't be used: this is always a full table scan?
- There are some edge cases when this would return the wrong thing? table,schema of ("one.two","three") would be the same as ("one","two.three")

Comment (niross) [reply dataworkspace/dataworkspace/datasets_db.py:49]: Sure, yes, this would require some work on the data flow side so happy to create a ticket to cover it (and include updating this in it). Currently we don't have any schema names with dots configured in data flow.

Comment (niross) [reply dataworkspace/dataworkspace/datasets_db.py:49]: Actually scratch that I got confused sorry. We can split the schema and table out but I it would mean running the query for every table separately. I went this route because we'd need to run this query for each table, so if there are queries with lots of tables and data cuts with lots of queries we would be running this quite a lot per page view.

Comment (niross) [reply dataworkspace/dataworkspace/datasets_db.py:49]: Unless... I guess I could build an or query. Let me have a think

Comment (michalc) [reply dataworkspace/dataworkspace/datasets_db.py:49]: How about something like...
```sql
SELECT
  DISTINCT ON (table_schema, table_name)
  table_schema,
  table_name,
  MAX(dataflow_swapped_tables_utc)
OVER
  (PARTITION BY table_schema, table_name)
FROM
  dataflow.metadata
WHERE
  (table_schema, table_name) IN (
    ('companieshouse', 'companies_match_ids'),
    ('dit', 'consent_service__current_consents')
  );
```

Comment (niross) [reply dataworkspace/dataworkspace/datasets_db.py:49]: Cheers, I am not sure if the partition is necessary here is it? Could I not achieve the same result with a group by? I've not used over/partition before so not sure what the benefit is.

```
SELECT
    table_schema,
    table_name,
    MAX(dataflow_swapped_tables_utc) AS run_date
FROM dataflow.metadata
WHERE (table_schema, table_name) IN (
    ('companieshouse', 'companies_match_ids'),
    ('dit', 'consent_service__current_consents')
)
GROUP BY (1, 2)
```

Comment (michalc) [reply dataworkspace/dataworkspace/datasets_db.py:49]: You're absolutely right: I was needlessly overcomplicating it

(Also: I didn't now you could group by by position in the select list...)

File: dataworkspace/dataworkspace/datasets_db.py [hunk 6]
```diff
@@ -18,18 +17,40 @@ def get_columns(database_name, schema=None, table=None, query=None):
...
-    sql = psycopg2.sql.SQL('SELECT * from {} WHERE false').format(source)
-    with connection.cursor() as cursor:
-        cursor.execute(sql)
-        return [c[0] for c in cursor.description]
+def get_tables_last_updated_date(database_name: str, tables: Tuple[Tuple[str, str]]):
+    """
+    Return the earliest of the last updated dates for a list of tables.
+    """
+    with connections[database_name].cursor() as cursor:
+        # Check the metadata table exists before we query it
```
Comment (michalc) [inline dataworkspace/dataworkspace/datasets_db.py:36]: Just realised... this does exist in production. Can we assume it exists? (together with creating it in any other environments?)

Comment (niross) [reply dataworkspace/dataworkspace/datasets_db.py:36]: Yeah was wavering on that myself. Happy to remove it
